src/architecture/merkledb.md

```diff
@@ -0,0 +1,207 @@
+
+- keys or values in case of list objects, sets or entries
+- key-value pairs in case of map objects.
+
+Such iterators use key ordering of the underlying key-value storage to
+determine the iteration order. Namely, keys are lexicographically ordered
+according to their binary serializations. This ordering coincides with that
+used in [RocksDB](storage.md).
+
+### MapIndex
+
+[`MapIndex`][map-index] implements a key-value store aka a map. It has
+the following functionality:
+
+- Get, set and remove value by key
+- Check if a specific key is present in the map
+- Iterate over the key-value pairs in the lexicographic key order
+- Iterate over keys in the lexicographic key order
+- Iterate over values in the lexicographic key order
+- Clear the map (i.e., remove all stored key-value pairs).
+
+### ListIndex
+
+[`ListIndex`][list-index] represents an array list.
+The following operations are supported:
+
+- Get and set a list item by index
+- Append an item to the list
+- Pop or poll the last item from the list
+- Get the list length
+- Check if the list is empty
+- Iterate over index-item pairs ordered by index
+- Insert a sequence of items from an iterator
+- Truncate the list to the specified length
+- Clear the list (i.e., remove all stored items from the list).
+
+`ListIndex` does not support inserting items in the middle of the
+list or removing items by index
+(although it is still possible to implement these operations manually).
+
+!!! summary "Implementation Details"
+    To support proper iteration, `ListIndex` saves its items with 8-byte
+    unsigned indices as keys. The indices are serialized in the big-endian form.
+
+### SparseListIndex
+
+[`SparseListIndex`][sparse-list-index] represents a `ListIndex` that may
+contain "gaps". It provides the possibility to delete items not only from
+the end of the list, but from any part thereof. Such deletions do not break
+the order of the indices inside the list.
+
+The remaining functionality of the `SparseListIndex` is the same as for
+[`ListIndex`](#ListIndex).
+
+### ValueSetIndex
+
+[`ValueSetIndex`][value-set-index] represents a hash set.
+The following operations are implemented:
+
+- Add and remove set items
+- Check if an item is already present using the item itself or its hash
+- Iterate over stored items in the lexicographic order of their hashes
+- Iterate over hashes of items in the lexicographic order
+- Clear the set (i.e., remove all items).
+
+The hash used in `ValueSetIndex` is calculated using the `object_hash()` method
+of the [`ObjectHash` trait][object-hash].
+
+!!! summary "Implementation Details"
+    Internally, `ValueSetIndex` uses hashes of items as keys,
+    and items themselves as corresponding values.
+
+### KeySetIndex
+
+[`KeySetIndex`][key-set-index] represents a set.
+The following procedures are implemented:
+
+- Add and remove set items
+- Check if a specific item is in the set
+- Iterate over items in the lexicographic order
```

Comment:
Shall we clarify that 

"in the lexicographic order of their serialized/binary representation"?